openlibrary/components/BarcodeScanner.vue

```diff
@@ -0,0 +1,423 @@
+            if (this.clearTimeout) clearTimeout(this.clearTimeout);
+            const drawingCtx = Quagga.canvas.ctx.overlay;
+            const drawingCanvas = Quagga.canvas.dom.overlay;
+            const canvasWidth = parseFloat(drawingCanvas.getAttribute('width'));
+            const canvasHeight = parseFloat(drawingCanvas.getAttribute('height'));
+            drawingCtx.clearRect(0, 0, canvasWidth, canvasHeight);
+        },
+
+        handleQuaggaDetected(result) {
+            const code = result.codeResult.code;
+            if (!this.isBarcodeISBN(code)) return;
+
+            this.submitISBNThrottled(code, Quagga.canvas.dom.image.toDataURL());
+        },
+
+        submitISBN(isbn, tentativeCoverUrl) {
+            if (isbn === this.lastISBN) return;
+            if (this.seenISBN.has(isbn)) return;
+
+            if (this.returnTo) {
+                location = this.returnTo.replace('$$$', isbn);
+            }
+            this.isbnList.unshift({isbn: isbn, cover: tentativeCoverUrl});
+            this.seenISBN.add(isbn);
+        },
+
+        isBarcodeISBN(code) {
+            return code.startsWith('97');
+        },
+
+        handleMultiISBN() {
+            if (location.href.includes('returnTo')) {
+                window.history.replaceState('', '', 'barcodescanner');
+                this.readMultiISBN = true;
+                this.returnTo = '';
+            } else {
+                window.history.replaceState('', '', 'barcodescanner?returnTo=/isbn/$$$');
+                this.readMultiISBN = false;
+                this.returnTo = new URLSearchParams(location.search).get('returnTo');
+            }
+        },
+
+        waitForShadowDOM() {
+            return new Promise(resolve => {
+                if (document.querySelector('#interactive')) {
+                    return resolve(document.querySelector('#interactive'));
+                }
+                if (document.querySelector('ol-barcode-scanner').shadowRoot.getElementById('interactive')) {
+                    return resolve(document.querySelector('ol-barcode-scanner').shadowRoot.getElementById('interactive'))
+                }
+                const observer = new MutationObserver(_mutations => {
+                    if (document.querySelector('ol-barcode-scanner').shadowRoot.getElementById('interactive')) {
+                        observer.disconnect();
+                        resolve(document.querySelector('ol-barcode-scanner').shadowRoot.getElementById('interactive'));
+                    }
+                })
+                observer.observe(document.body, {
+                    childList: true,
+                    subtree: true,
+                })
+            })
+        },
+    },
+    async mounted() {
+        this.targetInteractive = await this.waitForShadowDOM();
+        this.start();
+        // Disable Read Text ISBN button until camera permissions are granted
+        const observer = new MutationObserver(_mutations => {
```

Comment:
Instead of a mutation observer which can be a little complex/brittle/expensive, we can `await this.start()`. And have that method throw an error if Quagga fails to initialize -- which will indicate canvasInactive. (Note you might have to wrap the body of that function in a:

```js
return new Promise((res, rej) => {
     ...
});
```

So you can handle erroring inside the Quagga callback.

Reply:
I'm a little unclear on what this should look like, would this implementation be something along the lines of this?

```
start() {
  return new Promise((res, rej) => {
    Quagga.init({
      ...
    })
    if (!this.quaggaVideo) {
      rej();
    } else {
      Quagga.onProcessed(this.handleQuaggaProcessed.bind(this));
      Quagga.onDetected(this.handleQuaggaDetected.bind(this));
      res();
    }
```

and within async mounted()

```
await this.start()
  .then(() => {
    this.canvasInactive = false;
  }
````

Reply:
Very close! Something closer to this:

```js
start() {
  return new Promise((res, rej) => {
    Quagga.init({
      ...
    }, async (err) => {
        if (err) {
            rej(err);
            return}
        }
        ....
        res();
    });
```

This way we the promise is rejected if there's an error with Quagga (i.e. can't get access to video) and resolved if there isn't.

Then we can do:

```js
await this.start();
this.canvasInactive = false;
```

By using `await`, we no longer need to use `.then(` chaining. And if the promise is rejected, awaiting it will throw an error!

Reply:
(Also as a side note: Vue automatically `.bind`'s all the methods, so you can remove the `.bind(this)`!

Reply:
Thanks! I was pretty stumped on what I needed to do here